### PR TITLE
perf: speed up offset_by 2x for constant durations

### LINKED ISSUE #16722: Alternative method 10x faster than dt.offset_by()
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import pandas as pd
import polars as pl

num_rows = 1000000
utc_time = pd.date_range(start='2023-01-01', periods=num_rows, freq='ms')

# Create the DataFrame
df = pd.DataFrame({
    'UTC_Time': utc_time
})

df['UTC_Time'] = df['UTC_Time'].sort_values()

df = pl.DataFrame(df)

df = df.with_columns(pl.col("UTC_Time").dt.offset_by("5m").alias("Method1"))

df = df.with_columns(pl.from_epoch((pl.col("UTC_Time").dt.epoch(time_unit="ns") 
                                    + 5 * 60 * 1_000_000_000), time_unit="ns").alias("Method2"))

%timeit df.with_columns(pl.col("UTC_Time").dt.offset_by("5m"))                    
%timeit df.with_columns(pl.from_epoch((pl.col("UTC_Time").dt.epoch(time_unit="ns") + 5 * 60 * 1_000_000_000), time_unit="ns"))

```

### Log output

```shell
5.31 ms ± 337 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
534 µs ± 21.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

### Issue description

So I was wondering why dt.offset_by() was so slow and experimented with an alternative method. Turned out it was ~10x faster than the dt.offset_by() method. Would be nice to make dt.offset_by() equally as fast.

### Expected behavior

That dt.offset_by() becomes at least equally as fast as alternative method

### Installed versions

<details>

```
INSTALLED VERSIONS
------------------
commit                : bdc79c146c2e32f2cab629be240f01658cfb6cc2
python                : 3.12.2.final.0
python-bits           : 64
OS                    : Windows
OS-release            : 11
Version               : 10.0.22631
machine               : AMD64
processor             : Intel64 Family 6 Model 186 Stepping 2, GenuineIntel
byteorder             : little
LC_ALL                : None
LANG                  : en
LOCALE                : Norwegian Bokmål_Norway.1252

pandas                : 2.2.1
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.9.0.post0
setuptools            : 68.2.2
pip                   : 23.3.1
Cython                : 3.0.10
pytest                : None
hypothesis            : None
sphinx                : 7.2.6
blosc                 : None
feather               : None
xlsxwriter            : None
lxml.etree            : 5.2.2
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : 3.1.3
IPython               : 8.22.2
pandas_datareader     : 0.10.0
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.3
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : None
gcsfs                 : None
matplotlib            : 3.8.3
numba                 : 0.59.1
numexpr               : None
odfpy                 : None
openpyxl              : 3.1.2
pandas_gbq            : None
pyarrow               : 15.0.2
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : None
scipy                 : 1.12.0
sqlalchemy            : None
tables                : None
tabulate              : 0.9.0
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2024.1
qtpy                  : 2.4.1
pyqt5                 : None

pl.show_versions()
--------Version info---------
Polars:               0.20.31
Index type:           UInt32
Platform:             Windows-11-10.0.22631-SP0
Python:               3.12.2 | packaged by Anaconda, Inc. | (main, Feb 27 2024, 17:28:07) [MSC v.1916 64 bit (AMD64)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          3.0.0
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           3.8.3
nest_asyncio:         1.6.0
numpy:                1.26.4
openpyxl:             3.1.2
pandas:               2.2.1
pyarrow:              15.0.2
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
torch:                <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #16722  (timing becomes in line with the alternative method presented - I couldn't reproduce the 10x difference reported)

timing results available here https://www.kaggle.com/code/marcogorelli/polars-timing?scriptVersionId=181508652

---

~~I think this is fine but I didn't sleep too well so I'll check over it again tomorrow before marking as 'ready for review'~~

---

The gist of the change is: in the simple case, keep things simple and use `apply_values`

### CODE REVIEW

File: crates/polars-plan/src/dsl/function_expr/temporal.rs [hunk 1]
```diff
@@ -189,33 +189,54 @@ pub(super) fn datetime(
...
             Some(offset) => {
                 let offset = &Duration::parse(offset);
-                datetime
-                    .0
-                    .try_apply_nonnull_values_generic(|v| offset_fn(offset, v, time_zone))
+                if offset.is_constant_duration(datetime.time_zone().as_deref()) {
+                    // fastpath!
+                    match datetime.time_unit() {
+                        TimeUnit::Milliseconds => {
+                            Ok(datetime.0.apply_values(|v| v + offset.duration_ms()))
```
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply crates/polars-plan/src/dsl/function_expr/temporal.rs:202]: thanks @mcrumiller!

File: crates/polars-plan/src/dsl/function_expr/temporal.rs [hunk 2]
```diff
@@ -189,33 +189,53 @@ pub(super) fn datetime(
 fn apply_offsets_to_datetime(
     datetime: &Logical<DatetimeType, Int64Type>,
     offsets: &StringChunked,
-    offset_fn: fn(&Duration, i64, Option<&Tz>) -> PolarsResult<i64>,
     time_zone: Option<&Tz>,
 ) -> PolarsResult<Int64Chunked> {
```
Comment (ritchie46) [inline crates/polars-plan/src/dsl/function_expr/temporal.rs:193]: Just notice this now, but ideally the implementation should be in `polars-ops` or `polars-time` and then we only dispatch to those functions here. I'd like `polars-plan` to only be about the query planning/resolving.

File: crates/polars-plan/src/dsl/function_expr/temporal.rs [hunk 3]
```diff
@@ -189,33 +189,53 @@ pub(super) fn datetime(
...
-                    .0
-                    .try_apply_nonnull_values_generic(|v| offset_fn(offset, v, time_zone))
+                if offset.is_constant_duration(datetime.time_zone().as_deref()) {
+                    // fastpath!
+                    let duration = match datetime.time_unit() {
+                        TimeUnit::Milliseconds => offset.duration_ms(),
+                        TimeUnit::Microseconds => offset.duration_us(),
+                        TimeUnit::Nanoseconds => offset.duration_ns(),
+                    };
+                    if offset.negative() {
```
Comment (ritchie46) [inline crates/polars-plan/src/dsl/function_expr/temporal.rs:205]: To reduce binary bloat, I think we can make one code branch of this that does the `apply_values`.

```rust
if offset.negative() {
   duration = - duration;
}

Ok(datetime.0.apply_values(|v| v + duration))
```

**Comment by ritchie46** [comment]
Great hammering on those temporal performance! :muscle: Left some comments.

File: crates/polars-plan/src/dsl/function_expr/temporal.rs (hunk 2 above)
Comment (MarcoGorelli) [reply crates/polars-plan/src/dsl/function_expr/temporal.rs:193]: yup - ok if I do that separately?

**Review by ritchie46 — APPROVED** [review]

**Pull request merged**